polkadot/xcm/xcm-executor/src/assets.rs

```diff
@@ -132,34 +132,19 @@ impl AssetsInHolding {
 
 	/// Mutate `self` to contain all given `assets`, saturating if necessary.
 	///
-	/// NOTE: [`AssetsInHolding`] are always sorted, allowing us to optimize this function from
-	/// `O(n^2)` to `O(n)`.
+	/// NOTE: [`AssetsInHolding`] are always sorted
 	pub fn subsume_assets(&mut self, mut assets: AssetsInHolding) {
-		let mut f_iter = assets.fungible.iter_mut();
-		let mut g_iter = self.fungible.iter_mut();
-		if let (Some(mut f), Some(mut g)) = (f_iter.next(), g_iter.next()) {
-			loop {
-				if f.0 == g.0 {
-					// keys are equal. in this case, we add `self`'s balance for the asset onto
-					// `assets`, balance, knowing that the `append` operation which follows will
-					// clobber `self`'s value and only use `assets`'s.
-					(*f.1).saturating_accrue(*g.1);
-				}
-				if f.0 <= g.0 {
-					f = match f_iter.next() {
-						Some(x) => x,
-						None => break,
-					};
-				}
-				if f.0 >= g.0 {
-					g = match g_iter.next() {
-						Some(x) => x,
-						None => break,
-					};
-				}
-			}
+		// for fungibles, find matching fungibles and sum their amounts so we end-up having just
+		// single such fungible but with increased amount inside
+		for (k, v) in assets.fungible {
```

Comment:
Nit: Please let's use something like `(asset_id, asset_qty)` here. To make it easier to follow.